File: Coding/python/disjoint_union_sets.py
"""Disjoint set (union-find) and problems solved with it.

Idiomatic Python 3 port of java/DisjointUnionSets.java (original package code.ds).
"""


class DisjointUnionSets:
    def __init__(self, n):
        self.n = n
        self.rank = [0] * n
        self.parent = [0] * n
        self.make_set()

    # Create n sets, each containing a single item.
    def make_set(self):
        for i in range(self.n):
            self.parent[i] = i

    # Return the representative of x's set (with path compression).
    def find(self, x):
        if self.parent[x] != x:
            # x is not the representative of its set, so recurse on its parent
            # and hook x directly under the representative.
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    # Unite the set containing x with the set containing y (union by rank).
    def union(self, x, y):
        x_root = self.find(x)
        y_root = self.find(y)
        if x_root == y_root:
            return  # already in the same set
        if self.rank[x_root] < self.rank[y_root]:
            self.parent[x_root] = y_root
        elif self.rank[y_root] < self.rank[x_root]:
            self.parent[y_root] = x_root
        else:  # equal ranks: attach y under x and bump x's rank
            self.parent[y_root] = x_root
            self.rank[x_root] += 1

    # Count the number of islands in a 0/1 grid (8-connectivity).
    @staticmethod
    def count_islands(a):
        n = len(a)
        m = len(a[0])
        dus = DisjointUnionSets(n * m)

        # For each cell that is 1, union it with any neighbour that is also 1.
        for j in range(n):
            for k in range(m):
                if a[j][k] == 0:
                    continue
                if j + 1 < n and a[j + 1][k] == 1:
                    dus.union(j * m + k, (j + 1) * m + k)
                if j - 1 >= 0 and a[j - 1][k] == 1:
                    dus.union(j * m + k, (j - 1) * m + k)
                if k + 1 < m and a[j][k + 1] == 1:
                    dus.union(j * m + k, j * m + k + 1)
                if k - 1 >= 0 and a[j][k - 1] == 1:
                    dus.union(j * m + k, j * m + k - 1)
                if j + 1 < n and k + 1 < m and a[j + 1][k + 1] == 1:
                    dus.union(j * m + k, (j + 1) * m + k + 1)
                if j + 1 < n and k - 1 >= 0 and a[j + 1][k - 1] == 1:
                    dus.union(j * m + k, (j + 1) * m + k - 1)
                if j - 1 >= 0 and k + 1 < m and a[j - 1][k + 1] == 1:
                    dus.union(j * m + k, (j - 1) * m + k + 1)
                if j - 1 >= 0 and k - 1 >= 0 and a[j - 1][k - 1] == 1:
                    dus.union(j * m + k, (j - 1) * m + k - 1)

        c = [0] * (n * m)  # frequency of each set representative
        number_of_islands = 0
        for j in range(n):
            for k in range(m):
                if a[j][k] == 1:
                    x = dus.find(j * m + k)
                    if c[x] == 0:
                        number_of_islands += 1
                    c[x] += 1
        return number_of_islands

    # Given stacks of services, find which stacks can be deployed disjointly
    # (share no service in common).
    @staticmethod
    def get_stacks(input_, n):
        result = []
        seen = set()
        dus = DisjointUnionSets(n)
        dus.union(0, 1)
        dus.union(1, 2)
        dus.union(2, 0)
        for i in range(len(input_)):
            key_i = "Stack " + str(i + 1)
            if key_i not in seen:
                temp = [key_i]
                seen.add(key_i)
                for j in range(i + 1, len(input_)):
                    if dus.find(input_[i][0]) == dus.find(input_[j][0]):
                        key_j = "Stack " + str(j + 1)
                        temp.append(key_j)
                        seen.add(key_j)
                result.append(temp)
        return result

    # Bloomberg: number of friend circles in a char matrix ('x' marks a relation).
    @staticmethod
    def find_friend_circles(mat):
        m = len(mat)
        n = len(mat[0])
        seen = set()
        dus = DisjointUnionSets(m * n)
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 'x':
                    dus.union(i, j)
        for i in range(m):
            seen.add(dus.find(i))
        return len(seen)

    # DFS-based count of friend circles in an adjacency matrix.
    def find_circle_num(self, M):
        visited = [False] * len(M)  # visited[i]: is person i already counted
        count = 0
        for i in range(len(M)):
            if not visited[i]:
                self._dfs(M, visited, i)
                count += 1
        return count

    def _dfs(self, M, visited, person):
        for other in range(len(M)):
            if M[person][other] == 1 and not visited[other]:
                # Found an unvisited person in the current friend cycle.
                visited[other] = True
                self._dfs(M, visited, other)


if __name__ == "__main__":
    input_ = [[0, 1], [1, 2], [2, 1]]
    out = DisjointUnionSets.get_stacks(input_, 3)
    print("getStacks:", out)

    mat = [
        ['x', '.', '.', 'x'],
        ['.', 'x', '.', '.'],
        ['.', '.', 'x', 'x'],
        ['x', '.', 'x', 'x'],
    ]
    print("findFriendCircles:", DisjointUnionSets.find_friend_circles(mat))

    grid = [
        [1, 1, 0, 0, 0],
        [1, 1, 0, 0, 0],
        [0, 0, 0, 1, 1],
        [0, 0, 0, 1, 1],
    ]
    print("countIslands:", DisjointUnionSets.count_islands(grid))

    M = [
        [1, 1, 0],
        [1, 1, 0],
        [0, 0, 1],
    ]
    print("findCircleNum:", DisjointUnionSets(len(M)).find_circle_num(M))

File: Coding/golang/disjointunionsets/disjointunionsets.go
package main

import "fmt"

// DisjointUnionSets implements union-find with path compression + union by rank.
// find/union are ~O(α(n)) amortized.
type DisjointUnionSets struct {
	rank   []int
	parent []int
	n      int
}

func NewDisjointUnionSets(n int) *DisjointUnionSets {
	d := &DisjointUnionSets{
		rank:   make([]int, n),
		parent: make([]int, n),
		n:      n,
	}
	d.makeSet()
	return d
}

// makeSet creates n singleton sets (each element is its own parent).
func (d *DisjointUnionSets) makeSet() {
	for i := 0; i < d.n; i++ {
		d.parent[i] = i
	}
}

// find returns the representative of x's set (with path compression).
func (d *DisjointUnionSets) find(x int) int {
	if d.parent[x] != x {
		d.parent[x] = d.find(d.parent[x])
	}
	return d.parent[x]
}

// union merges the sets containing x and y (by rank).
func (d *DisjointUnionSets) union(x, y int) {
	xRoot, yRoot := d.find(x), d.find(y)
	if xRoot == yRoot {
		return
	}
	if d.rank[xRoot] < d.rank[yRoot] {
		d.parent[xRoot] = yRoot
	} else if d.rank[yRoot] < d.rank[xRoot] {
		d.parent[yRoot] = xRoot
	} else {
		d.parent[yRoot] = xRoot
		d.rank[xRoot]++
	}
}

// countIslands counts connected groups of 1s using 8-neighbour connectivity.
func countIslands(a [][]int) int {
	n := len(a)
	m := len(a[0])
	dus := NewDisjointUnionSets(n * m)

	// Check each cell's 8 neighbours and union them if both are land (1).
	for j := 0; j < n; j++ {
		for k := 0; k < m; k++ {
			if a[j][k] == 0 {
				continue
			}
			if j+1 < n && a[j+1][k] == 1 {
				dus.union(j*m+k, (j+1)*m+k)
			}
			if j-1 >= 0 && a[j-1][k] == 1 {
				dus.union(j*m+k, (j-1)*m+k)
			}
			if k+1 < m && a[j][k+1] == 1 {
				dus.union(j*m+k, j*m+k+1)
			}
			if k-1 >= 0 && a[j][k-1] == 1 {
				dus.union(j*m+k, j*m+k-1)
			}
			if j+1 < n && k+1 < m && a[j+1][k+1] == 1 {
				dus.union(j*m+k, (j+1)*m+k+1)
			}
			if j+1 < n && k-1 >= 0 && a[j+1][k-1] == 1 {
				dus.union(j*m+k, (j+1)*m+k-1)
			}
			if j-1 >= 0 && k+1 < m && a[j-1][k+1] == 1 {
				dus.union(j*m+k, (j-1)*m+k+1)
			}
			if j-1 >= 0 && k-1 >= 0 && a[j-1][k-1] == 1 {
				dus.union(j*m+k, (j-1)*m+k-1)
			}
		}
	}

	// Count distinct representatives among land cells.
	c := make([]int, n*m)
	numberOfIslands := 0
	for j := 0; j < n; j++ {
		for k := 0; k < m; k++ {
			if a[j][k] == 1 {
				x := dus.find(j*m + k)
				if c[x] == 0 {
					numberOfIslands++
				}
				c[x]++
			}
		}
	}
	return numberOfIslands
}

// getStacks groups "service stacks" that share a common service into disjoint sets.
func getStacks(input [][]int, n int) [][]string {
	result := make([][]string, 0)
	set := make(map[string]bool)
	dus := NewDisjointUnionSets(n)
	dus.union(0, 1)
	dus.union(1, 2)
	dus.union(2, 0)
	for i := 0; i < len(input); i++ {
		key := fmt.Sprintf("Stack %d", i+1)
		if !set[key] {
			temp := []string{key}
			set[key] = true
			for j := i + 1; j < len(input); j++ {
				if dus.find(input[i][0]) == dus.find(input[j][0]) {
					sj := fmt.Sprintf("Stack %d", j+1)
					temp = append(temp, sj)
					set[sj] = true
				}
			}
			result = append(result, temp)
		}
	}
	return result
}

// findFriendCircles counts friend circles in a char matrix ('x' links row i and col j).
func findFriendCircles(mat [][]byte) int {
	m := len(mat)
	n := len(mat[0])
	set := make(map[int]bool)
	dus := NewDisjointUnionSets(m * n)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if mat[i][j] == 'x' {
				dus.union(i, j)
			}
		}
	}
	for i := 0; i < m; i++ {
		set[dus.find(i)] = true
	}
	return len(set)
}

// findCircleNum counts friend circles via DFS over an adjacency matrix.
func (d *DisjointUnionSets) findCircleNum(m [][]int) int {
	visited := make([]bool, len(m)) // visited[i] => person i already grouped
	count := 0
	for i := 0; i < len(m); i++ {
		if !visited[i] {
			d.dfs(m, visited, i)
			count++
		}
	}
	return count
}

func (d *DisjointUnionSets) dfs(m [][]int, visited []bool, person int) {
	for other := 0; other < len(m); other++ {
		if m[person][other] == 1 && !visited[other] {
			visited[other] = true
			d.dfs(m, visited, other)
		}
	}
}

func main() {
	input := [][]int{{0, 1}, {1, 2}, {2, 1}}
	out := getStacks(input, 3)
	fmt.Println("Disjoint stacks:", out)

	mat := [][]byte{
		{'x', '.', '.', 'x'},
		{'.', 'x', '.', '.'},
		{'.', '.', 'x', 'x'},
		{'x', '.', 'x', 'x'},
	}
	fmt.Println(findFriendCircles(mat))
}
